charmcraft/models/project.py

```diff
@@ -345,38 +347,28 @@ def get_build_plan(self) -> list[models.BuildInfo]:
 
         if self.platforms is None:
             raise CraftError("Must define at least one platform.")
-        build_infos = []
-        for platform_name, platform in self.platforms.items():
-            if platform is None:
-                if platform_name not in const.SUPPORTED_ARCHITECTURES:
-                    raise CraftError(
-                        f"Invalid platform {platform_name}.",
-                        details="A platform name must either be a valid architecture name or the "
-                        "platform must specify one or more build-on and build-for architectures.",
-                    )
-                build_infos.append(
-                    models.BuildInfo(
-                        platform_name,
-                        build_on=platform_name,
-                        build_for=platform_name,
-                        base=base,
-                    )
-                )
-            else:
-                # TODO: this should go to craft-platforms, so silence mypy for now.
-                for build_on in platform.build_on:  # type: ignore[union-attr]
-                    build_infos.extend(
-                        [
-                            models.BuildInfo(
-                                platform_name,
-                                build_on=str(build_on),
-                                build_for=str(build_for),
-                                base=base,
-                            )
-                            for build_for in platform.build_for  # type: ignore[union-attr]
-                        ]
-                    )
-        return build_infos
+        platforms = cast(
+            # https://github.com/canonical/craft-platforms/issues/43
+            craft_platforms.Platforms,  # pyright: ignore[reportPrivateImportUsage]
+            {
+                name: (platform.marshal() if platform else None)
```

Comment:
this is just me rambling but it bothers me a bit that we have to do this 'round-trip' where we unmarshal the project, which includes validation (necessary) and possibly transformation (dangerous), and then re-marshal to pass to craft-platforms. This makes the assembling of the build plan a *little* different between the application and systems that just use craft-platforms (like launchpad), and makes us still need "all" of this code here.

I don't have any good suggestions to address this though; maybe the base BuildPlanner class should retain the original values of base, build-base and platforms somewhere?